dlib/pixel.h

```diff
@@ -107,6 +107,43 @@ namespace dlib
                 - is_unsigned is true if min() == 0 and false otherwise
     !*/
 
+// ----------------------------------------------------------------------------------------
+
+    namespace details
+    {
+        template<typename...T>
+        using void_t = void;
+
+        template<class Pixel, class = void>
+        struct is_pixel_type : std::false_type{};
+
+        template<class Pixel>
+        struct is_pixel_type<Pixel, void_t<decltype(pixel_traits<Pixel>::grayscale),
+                                           decltype(pixel_traits<Pixel>::rgb),
+                                           decltype(pixel_traits<Pixel>::rgb_alpha),
+                                           decltype(pixel_traits<Pixel>::hsi),
+                                           decltype(pixel_traits<Pixel>::lab),
+                                           decltype(pixel_traits<Pixel>::has_alpha),
+                                           decltype(pixel_traits<Pixel>::bgr_layout),
+                                           decltype(pixel_traits<Pixel>::num),
+                                           decltype(pixel_traits<Pixel>::is_unsigned)>> : std::true_type{};
```

Comment:
Also the `details` namespace isn't needed.  Just let `is_pixel_type` be in `dlib` since there isn't anything else going on here (aside from `void_t` but totally move that to algs.h).

Reply:
`void_t` is already in `dlib/type_traits.h`. I'll keep it there, I think it's more appropriate, as that header contains backports of later c++ standards and currently `std::void_t` lives in `<type_traits>` in C++17.

Reply:
Yeah my objective was to make that type trait more precise and follow the documentation for what a pixel trait should be to the letter. Then it could be used as an emulated concept using `enable_if_t<pixel_trait<>,bool>` and later as a proper concept. I didn't think it was right to have a type trait checker if it didn't check the type trait was documentation conforming. 
I guess we can do what you say but you could have an empty type trait and it would still work. I'm not convinced that's right.

Reply:
Well, it will fail though when something tries to use it.  And at that point the compiler error will mention the name of the missing field.  So it will be a more readable error message.  That's all I'm saying.  But if you wanted to put a unit test that checked that all the dlib pixel types have correct traits that's cool.  Or did something to make the error message about a bad traits type even better than whatever the compiler normally emits that's cool too.  Although I think the normal error message for that kind of thing is fine as is.  Might be hard to make it more clear. IDK.

Reply:
Ok. We'll just leave it as it was, roughly. 
I've added `is_complete_type<T>` type trait and now you can just write:
```
template <class Pixel>
using is_pixel_type = is_complete_type<pixel_traits<Pixel>>;
```
That's really clear. It says exactly what `is_pixel_type` is doing, i.e. a `pixel_trait` specialisation exists and also behaves in the same way as before.